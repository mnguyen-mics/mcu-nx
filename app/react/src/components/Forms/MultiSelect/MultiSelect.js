import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Icon, Dropdown, Menu, Button } from 'antd';
import { FormattedMessage } from 'react-intl';

class MultiSelect extends Component {

  constructor(props) {
    super(props);
    this.buildMenuItems = this.buildMenuItems.bind(this);
    this.handleVisibleChange = this.handleVisibleChange.bind(this);
    this.onMenuClick = this.onMenuClick.bind(this);
    this.state = {
      overlayVisible: false,
      selectedItems: this.props.menuItems.selectedItems
    };
  }


  render() {

    const {
      displayElement,
      buttonClass
    } = this.props;

    const {
      overlayVisible
    } = this.state;

    const menu = this.buildMenuItems();

    return (
      <Dropdown overlay={menu} trigger={['click']} onVisibleChange={this.handleVisibleChange} visible={overlayVisible}>
        <Button className={buttonClass}>
          { displayElement }
        </Button>
      </Dropdown>
    );

  }

  onMenuClick(item) {

    const {
      name,
      menuItems: {
        handleMenuClick
      }
    } = this.props;

    const {
      selectedItems
    } = this.state;

    // Add or remove item on selectedItems
    let newArray = [];
    const index = selectedItems.findIndex(selectedItem => selectedItem.value === item.value);
    if (index !== -1) {
      newArray = [
        ...selectedItems.slice(0, index),
        ...selectedItems.slice(index + 1)
      ];
    } else {
      newArray = [
        ...selectedItems,
        item
      ];
    }
    this.setState({
      selectedItems: newArray
    });

    if (typeof handleMenuClick === 'function') {
      handleMenuClick({
        [name]: newArray
      });
    }

  }

  buildMenuItems() {
    const {
      menuItems: {
        items
      }
    } = this.props;

    const {
      selectedItems
    } = this.state;

    const getItem = value => items.find(item => item.value === value);

    return (
      <Menu onClick={value => this.onMenuClick(getItem(value.key))}>
        { items.map(item => {
          const isItemSelected = selectedItems.find(selectedItem => selectedItem.value === item.value);
          return (
            <Menu.Item key={item.value}>
              {isItemSelected && (<Icon type="check" />)}
              <span><FormattedMessage id={item.key} /></span>
            </Menu.Item>
          );
        })}
      </Menu>
    );

  }

  setVisibility(isVisible) {
    this.setState({
      overlayVisible: isVisible
    });
  }

  handleVisibleChange(isVisible) {
    this.setVisibility(isVisible);
    if (!isVisible) {
      this.props.onCloseMenu(this.state.selectedItems);
    }
  }

  isChecked(value) {
    let isChecked = false;
    this.buildFilterItems().forEach((item) => {
      if (item.value.toLowerCase() === value.toLowerCase()) {
        isChecked = true;
      }
    });
    return isChecked;
  }

}

MultiSelect.defaultProps = {
  buttonClass: ''
};

MultiSelect.propTypes = {
  name: PropTypes.string.isRequired,
  displayElement: PropTypes.element.isRequired,
  onCloseMenu: PropTypes.func,
  menuItems: PropTypes.shape({
    handleMenuClick: PropTypes.func,
    selectedItems: PropTypes.arrayOf(PropTypes.shape({
      key: PropTypes.string,
      value: PropTypes.string,
    })),
    items: PropTypes.arrayOf(PropTypes.shape({
      key: PropTypes.string,
      value: PropTypes.string,
    }))
  }).isRequired,
  buttonClass: PropTypes.string,
};

MultiSelect.defaultProps = {
  onCloseMenu: () => { }
};

export default MultiSelect;
