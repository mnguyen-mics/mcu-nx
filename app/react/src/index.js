import React from 'react';
import { render } from 'react-dom';
import { Provider } from 'react-redux';

import configureStore from './store';
import { Navigator } from './containers/Navigator';

const store = configureStore();

const micsProvider = (
  <Provider store={store}>
    <Navigator />
  </Provider>
);

render(
  micsProvider,
  document.getElementById('mcs-react-app') // eslint-disable-line no-undef
);
