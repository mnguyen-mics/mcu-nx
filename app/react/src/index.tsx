/* eslint-disable */
import React from 'react';
import { render } from 'react-dom';

import App from './App';

declare global {
  interface Window {
    require: any;
    mainRev: any;
  }
}

render(<App />, document.getElementById('mcs-react-app'), () => {
  window.require([window.mainRev], () => {
    //
  });
});
