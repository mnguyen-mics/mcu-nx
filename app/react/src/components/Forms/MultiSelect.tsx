import * as React from 'react';
import { Icon, Dropdown, Menu, Button } from 'antd';

interface MultiSelectProps {
  name: string;
  displayElement: JSX.Element;
  onCloseMenu?: Function;
  // menuItems: {
  //   handleMenuClick?: any;
  //   selectedItems?: [{
  //     key?: string;
  //     value?: string;
  //   }];
  //   items: [{
  //     key?: string;
  //     value?: string;
  //   }];
  // };
  menuItems?: any;
  selectedItems?: any;
  buttonClass: any;
  key: string;
}

interface MultiSelectState {
  selectedItems: any;
  overlayVisible: boolean;
}

class MultiSelect extends React.Component<MultiSelectProps, MultiSelectState> {

  static defaultprops = {
    buttonClass: '',
    onCloseMenu: () => {},
  }

  constructor(props) {
    super(props);

    this.state = {
      overlayVisible: false,
      selectedItems: this.props.menuItems.selectedItems,
    };
  }

  buildMenuItems = () => {
    const { menuItems: { items } } = this.props;
    const { selectedItems } = this.state;
    const getItem = value => items.find(item => item.value === value);

    return (
      <Menu onClick={value => this.onMenuClick(getItem(value.key))}>
        { items.map(item => {
          const isItemSelected = selectedItems.find(selectedItem => selectedItem.value === item.value);
          return (
            <Menu.Item key={item.value}>
              {isItemSelected && (<Icon type="check" />)}
              <span>{item.key}</span>
            </Menu.Item>
          );
        })}
      </Menu>
    );

  }

  handleVisibleChange = (isVisible) => {
    this.setVisibility(isVisible);

    if (!isVisible) {
      this.props.onCloseMenu(this.state.selectedItems);
    }
  }

  isChecked(value) {
    let isChecked = false;

    // this.buildFilterItems().forEach((item) => {
    //   if (item.value.toLowerCase() === value.toLowerCase()) {
    //     isChecked = true;
    //   }
    // });

    return isChecked;
  }

  onMenuClick = (item) => {
    const { name, menuItems: { handleMenuClick } } = this.props;
    const { selectedItems } = this.state;

    // Add or remove item on selectedItems
    let newArray = [];
    const index = selectedItems.findIndex(selectedItem => selectedItem.value === item.value);

    if (index !== -1) {
      newArray = [
        ...selectedItems.slice(0, index),
        ...selectedItems.slice(index + 1),
      ];
    } else {
      newArray = [
        ...selectedItems,
        item,
      ];
    }

    this.setState({ selectedItems: newArray });

    if (typeof handleMenuClick === 'function') {
      handleMenuClick({
        [name]: newArray,
      });
    }
  }

  setVisibility(isVisible) {
    this.setState({
      overlayVisible: isVisible,
    });
  }

  render() {
    const { displayElement, buttonClass } = this.props;
    const { overlayVisible } = this.state;
    const menu = this.buildMenuItems();

    return (
      <Dropdown
        overlay={menu}
        trigger={['click']}
        onVisibleChange={this.handleVisibleChange}
        visible={overlayVisible}
      >
        <Button className={buttonClass}>
          { displayElement }
        </Button>
      </Dropdown>
    );

  }
}

export default MultiSelect;

