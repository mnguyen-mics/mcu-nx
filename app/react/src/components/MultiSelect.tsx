import * as React from 'react';
import {Icon, Dropdown, Menu, Button} from 'antd';
import {ClickParam} from 'antd/lib/menu';
import {ReactNode} from 'react-redux';

export interface MultiSelectProps<T> {
  displayElement: JSX.Element;

  items: T[];
  selectedItems: T[];

  display?: (t: T) => ReactNode;
  getKey: (t: T) => string;

  handleItemClick?: (item: T) => void;
  handleMenuClick?: (selectedItems: T[]) => void;
  onCloseMenu?: (selectedItems: T[]) => void;

  singleSelectOnly?: boolean;
  buttonClass?: string;
}

export interface MultiSelectState<T> {
  selectedItems: T[];
  overlayVisible: boolean;
}

class MultiSelect<T> extends React.Component<MultiSelectProps<T>, MultiSelectState<T>> {

  static defaultProps: Partial<MultiSelectProps<any>> = {
    buttonClass: '',
    display: t => t.toString(),
  };

  state = {
    overlayVisible: false,
    selectedItems: this.props.selectedItems,
  };

  buildMenuItems = () => {
    const {items, display, getKey} = this.props;
    const {selectedItems} = this.state;

    return (
      <Menu onClick={this.onMenuClick}>
        {items.map(item => {
          const isItemSelected = selectedItems.findIndex(selectedItem => getKey(selectedItem) === getKey(item)) !== -1;
          return (
            <Menu.Item key={getKey(item)}>
              {isItemSelected && (<Icon type="check"/>)}
              <span>{display!(item)}</span>
            </Menu.Item>
          );
        })}
      </Menu>
    );
  }

  handleVisibleChange = (isVisible: boolean) => {
    this.setVisibility(isVisible);

    if (!isVisible && this.props.onCloseMenu) {
      this.props.onCloseMenu(this.state.selectedItems);
    }
  }

  onMenuClick = (param: ClickParam) => {
    const {items, handleMenuClick, getKey, handleItemClick, singleSelectOnly} = this.props;
    const {selectedItems} = this.state;

    const clickedItem = items.find(item => getKey(item) === param.key);

    // Add or remove item on selectedItems
    let newArray: T[] = [];
    const index = selectedItems!.findIndex(selectedItem => getKey(selectedItem) === getKey(clickedItem!));

    if (index !== -1) {
      newArray = singleSelectOnly ? [] :
        [
          ...selectedItems.slice(0, index),
          ...selectedItems.slice(index + 1),
        ];
    } else {
      newArray = singleSelectOnly ? [clickedItem!] :
        [
          ...selectedItems,
          clickedItem!,
        ];
    }

    this.setState({selectedItems: newArray});

    if (handleMenuClick) {
      handleMenuClick(newArray);
    }
    if (handleItemClick) {
      handleItemClick(clickedItem!);
    }

  }

  setVisibility(isVisible: boolean) {
    this.setState({
      overlayVisible: isVisible,
    });
  }

  render() {
    const {displayElement, buttonClass} = this.props;
    const {overlayVisible} = this.state;
    const menu = this.buildMenuItems();

    return (
      <Dropdown
        overlay={menu}
        trigger={['click']}
        onVisibleChange={this.handleVisibleChange}
        visible={overlayVisible}
      >
        <Button className={buttonClass}>
          {displayElement}
        </Button>
      </Dropdown>
    );

  }
}

export default MultiSelect;
