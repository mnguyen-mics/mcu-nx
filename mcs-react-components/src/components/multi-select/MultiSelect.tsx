import * as React from 'react';
import { CheckOutlined } from '@ant-design/icons';
import { Menu, Button } from 'antd';
import { MenuInfo } from '../../../node_modules/rc-menu/lib/interface';
import SubMenu from 'antd/lib/menu/SubMenu';
import { Dropdown } from '../popupContainer/PopupContainer';

export interface MultiSelectProps<T> {
  displayElement: JSX.Element;
  items: T[];
  subItems?: T[];
  subItemsTitle?: string;
  selectedItems: T[];
  display?: (t: T) => React.ReactNode;
  getKey: (t: T) => string;
  handleItemClick?: (item: T) => void;
  handleMenuClick?: (selectedItems: T[]) => void;
  onCloseMenu?: (selectedItems: T[]) => void;
  singleSelectOnly?: boolean;
  buttonClass?: string;
}

export interface MultiSelectState<T> {
  overlayVisible: boolean;
}

class MultiSelect<T> extends React.Component<MultiSelectProps<T>, MultiSelectState<T>> {
  static defaultProps: Partial<MultiSelectProps<any>> = {
    buttonClass: '',
    display: t => t.toString(),
  };

  state = {
    overlayVisible: false,
  };

  buildMenuItems = () => {
    const { items, display, getKey, subItems, subItemsTitle } = this.props;
    const selectedItems = this.props.selectedItems;
    const renderItems = (elements: T[]) => {
      return elements.map(element => {
        const isItemSelected =
          selectedItems.findIndex(selectedItem => getKey(selectedItem) === getKey(element)) !== -1;
        return (
          <Menu.Item key={getKey(element)}>
            {isItemSelected && <CheckOutlined />}
            <span>{display!(element)}</span>
          </Menu.Item>
        );
      });
    };
    return (
      <Menu onClick={this.onMenuClick}>
        {renderItems(items)}
        {subItems && subItemsTitle && (
          <SubMenu title={subItemsTitle}>{renderItems(subItems)}</SubMenu>
        )}
      </Menu>
    );
  };

  handleVisibleChange = (isVisible: boolean) => {
    this.setVisibility(isVisible);

    if (!isVisible && this.props.onCloseMenu) {
      this.props.onCloseMenu(this.props.selectedItems);
    }
  };

  onMenuClick = (param: MenuInfo) => {
    const { items, handleMenuClick, getKey, handleItemClick, singleSelectOnly, subItems } =
      this.props;
    const { selectedItems } = this.props;

    const allItems = subItems ? items.concat(subItems) : items;

    const clickedItem = allItems.find(item => getKey(item) === param.key);

    // Add or remove item on selectedItems
    let newArray: T[] = [];
    const index = selectedItems!.findIndex(
      selectedItem => getKey(selectedItem) === getKey(clickedItem!),
    );

    if (index !== -1) {
      newArray = singleSelectOnly
        ? []
        : [...selectedItems.slice(0, index), ...selectedItems.slice(index + 1)];
    } else {
      newArray = singleSelectOnly ? [clickedItem!] : [...selectedItems, clickedItem!];
    }

    if (handleMenuClick) {
      handleMenuClick(newArray);
    }
    if (handleItemClick) {
      handleItemClick(clickedItem!);
    }
  };

  setVisibility(isVisible: boolean) {
    this.setState({
      overlayVisible: isVisible,
    });
  }

  render() {
    const { displayElement, buttonClass } = this.props;
    const { overlayVisible } = this.state;
    const menu = this.buildMenuItems();

    return (
      <Dropdown
        overlay={menu}
        trigger={['click']}
        onVisibleChange={this.handleVisibleChange}
        visible={overlayVisible}
      >
        <Button className={buttonClass}>{displayElement}</Button>
      </Dropdown>
    );
  }
}

export default MultiSelect;
