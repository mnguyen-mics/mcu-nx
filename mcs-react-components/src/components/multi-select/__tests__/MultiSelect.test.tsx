import * as React from 'react';
import { MemoryRouter } from 'react-router';
import { IntlProvider } from 'react-intl';
import { render } from 'enzyme';
import MultiSelect, { MultiSelectProps } from '../MultiSelect';

jest.mock('cuid', () => () => '123');

it('renders the multi select', () => {
  const props: MultiSelectProps<string> = {
    displayElement: <div>Click To Test</div>,
    items: ['Willy Denzey', 'Matt Huston'],
    subItems: ['Billy Crawford', 'TragÃ©die'],
    subItemsTitle: '... mais aussi',
    selectedItems: [],
    getKey: (a: string) => a,
    handleItemClick: (a: string) => console.log(a),
    handleMenuClick: (selectedItems: string[]) =>
      selectedItems.forEach(s => console.log('> clicked on: ' + s)),
    onCloseMenu: (selectedItems: string[]) =>
      selectedItems.forEach(s => console.log('> closing: ' + s)),
    singleSelectOnly: false,
    buttonClass: 'test-class',
  };

  const component = render(
    <IntlProvider locale="en">
      <MemoryRouter>
        <MultiSelect {...props} />
      </MemoryRouter>
    </IntlProvider>,
  );
  expect(component).toMatchSnapshot();
});
