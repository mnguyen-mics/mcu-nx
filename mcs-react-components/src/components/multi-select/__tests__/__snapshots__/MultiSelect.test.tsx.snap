// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the multi select 1`] = `
initialize {
  "0": Node {
    "attribs": Object {
      "id": "123",
    },
    "children": Array [
      Node {
        "attribs": Object {
          "class": "ant-btn ant-dropdown-trigger test-class",
          "type": "button",
        },
        "children": Array [
          Node {
            "attribs": Object {},
            "children": Array [
              Node {
                "data": "Click To Test",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {},
            "x-attribsPrefix": Object {},
          },
        ],
        "name": "button",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "type": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "type": undefined,
        },
      },
    ],
    "name": "span",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "id": undefined,
    },
    "x-attribsPrefix": Object {
      "id": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
